presto-raptor/src/main/java/com/facebook/presto/raptor/storage/BucketBalancer.java

```diff
@@ -0,0 +1,285 @@
+                    .max(Integer::compare)
+                    .get();
+
+            int targetMin = (int) Math.floor((existingAllocations.size() * 1.0) / activeNodes.size());
+            int targetMax = (int) Math.ceil((existingAllocations.size() * 1.0) / activeNodes.size());
+
+            log.info("Distribution %s: Current bucket skew: min %s, max %s. Target bucket skew: min %s, max %s", distribution.getId(), currentMin, currentMax, targetMin, targetMax);
+
+            for (String source : ImmutableSet.copyOf(allocationCounts)) {
+                for (BucketAssignment existingAssignment : ImmutableList.copyOf(existingAllocations.get(source))) {
+                    if (activeNodes.contains(source) && allocationCounts.count(source) <= targetMin) {
+                        break;
+                    }
+
+                    // identify nodes with bucket counts lower than the computed target, and greedily select from this set based on projected disk utilization.
+                    String target = activeNodes.stream()
+                            .filter(candidate -> !candidate.equals(source) && allocationCounts.count(candidate) < targetMax)
+                            .sorted((a, b) -> Integer.compare(allocationCounts.count(a), allocationCounts.count(b)))
+                            .min((a, b) -> Double.compare(allocationBytes.get(a), allocationBytes.get(b)))
+                            .orElseThrow(() -> new IllegalStateException("unable to find target for rebalancing"));
+
+                    // don't move bucket if it doesn't reduce imbalance
+                    if (activeNodes.contains(source) && (
+                            allocationCounts.count(source) == targetMax && allocationCounts.count(target) == targetMin &&
+                            (allocationBytes.get(target) + bucketSize) >= allocationBytes.get(source))) {
+                        break;
+                    }
+
+                    allocationCounts.remove(source);
+                    allocationCounts.add(target);
+                    allocationBytes.compute(source, (k, v) -> v - bucketSize);
+                    allocationBytes.compute(target, (k, v) -> v + bucketSize);
+
+                    sourceToAllocationChanges.put(
+                            existingAssignment.getNodeIdentifier(),
+                            assignment(existingAssignment.getDistributionId(), existingAssignment.getBucketNumber(), target));
+                }
+            }
+        }
+
+        // perform assignments in decreasing order of source node load
+        List<String> sourceNodes = sourceToAllocationChanges.asMap().entrySet().stream()
+                .sorted((a, b) -> Integer.compare(b.getValue().size(), a.getValue().size()))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toList());
+
+        log.info("Moving %s buckets...", sourceToAllocationChanges.size());
+
+        for (String source : sourceNodes) {
```

Comment:
It would help testing to have a separate method that computes the new assignments and a method that effectively update these assignments. This would for example allow to inspect the updates against the original assignments.